wrappers/kmip/kmip.go

```diff
@@ -0,0 +1,537 @@
+	if err != nil {
+		return nil, err
+	}
+	defer client.Close()
+
+	keyId := kms.keyId
+
+	// When using RSA, some KMIP servers may accept encryption using the private key ID, some others only accept the public key ID.
+	// Let's play it safe and always use the public key for encryption.
+	// The public key id is looked-up during initialization. It's empty when running in symmetric mode.
+	if kms.pubKeyId != "" {
+		keyId = kms.pubKeyId
+	}
+
+	encryptRespPayload, err := client.Encrypt(keyId).
+		WithIvCounterNonce(iv).
+		WithCryptographicParameters(kms.cryptoParams).
+		Data(plainDEK).
+		ExecContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Prepend the IV (if any) to the ciphertext
+	ciphertext := append(iv, encryptRespPayload.Data...)
+	// Append the encryption tag (if any) to the ciphertext. Depending on the KMIP version and the implementation,
+	// sometimes the tag is already appended to th cipher text, sometimes it's in the AuthenticatedEncryptionTag field.
+	// When using RSA-OAEP, this is a NOOP as AuthenticatedEncryptionTag is empty.
+	ciphertext = append(ciphertext, encryptRespPayload.AuthenticatedEncryptionTag...)
+	return ciphertext, nil
+}
+
+// decryptDEK uses the KMIP decrypt operation  to decrypt the DEK.
+func (kms *Wrapper) decryptDEK(ctx context.Context, keyId string, cparams kmip.CryptographicParameters, encryptedDEK []byte) ([]byte, error) {
+	nonceSize := int(kms.cryptoParams.IVLength)
+	tagLen := int(kms.cryptoParams.TagLength)
+	if len(encryptedDEK) < nonceSize+tagLen {
+		return nil, errors.New("invalid encrypted DEK")
+	}
+
+	client, err := kms.connect(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer client.Close()
+
+	var iv []byte
+	if nonceSize > 0 {
+		iv = encryptedDEK[:nonceSize]
+		encryptedDEK = encryptedDEK[nonceSize:]
+	}
+
+	reqBuilder := client.Decrypt(keyId).
+		WithIvCounterNonce(iv).
+		WithCryptographicParameters(cparams)
+
+	// Starting from kmip 1.4, GCM authentication tag is handled separately
+	// in a dedicated field. Prior 1.4, it's appended at the end of the ciphertext.
+	// (In reality, it mainly depends on the vendors)
+	if tagLen > 0 && ttlv.CompareVersions(client.Version(), kmip.V1_4) >= 0 {
+		reqBuilder = reqBuilder.WithAuthTag(encryptedDEK[len(encryptedDEK)-tagLen:])
+		encryptedDEK = encryptedDEK[:len(encryptedDEK)-tagLen]
+	}
+
+	decryptRespPayload, err := reqBuilder.
+		Data(encryptedDEK).
+		ExecContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return decryptRespPayload.Data, nil
+}
+
+// connect to the kmip endpoint, perform TLS and KMIP handshakes.
+func (kms *Wrapper) connect(ctx context.Context) (*kmipclient.Client, error) {
+	var conn *kmipclient.Client
+	var err error
+	if kms.cachedClient != nil {
+		// Cloning the cached client establish a new connection, reusing previously set parameters.
+		// It also skips protocol version negociation, reusing previously negociated version.
+		conn, err = kms.cachedClient.CloneCtx(ctx)
```

Comment:
Just making sure, is `cachedClient.CloneCtx(...)` thread safe?

Reply:
Yes it is. It creates a new fresh connection, and clones a few immutable fields